### PR TITLE
[WIP] cuDF backend for FeatureTools

### PR DESCRIPTION
**Pull request by VibhuJawa**
### [WIP] cuDF backend for FeatureTools
This PR adds cuDF backend and resolves https://github.com/alteryx/featuretools/issues/906.   

This still very much a  **WIP** but I want to open this up to get initial reviews. 
#### Tentative Performence
```python
Feature tools(Deep Feature Sythesis) speed up: cudf-dfs=  5.58 s vs pandas=156.99 (28 x)
E2E Overall(reading+dfs, writing):  21.4 s vs 3min 31s (9.85 x) 
```
#### Benchmark Caveat
Please note that this is not a benchmark that is very well suited for `cudf` as we have really wide frames here (`356_255` x `1564` ).  A more ideal benchmark will be one that is taller and less wide but even on this nonideal benchmark, the performance gain is commendable. 

### Bugs with cudf+Featuretools  (Blocking for mentioned features) 

This will serve as the running list of problems we need to resolve for all test cases to pass for the integration. 

- [ ] Related to **Datetime Variables**

    Blocks feature:  [Training window ](https://docs.featuretools.com/en/v0.17.0/automated_feature_engineering/handling_time.html#training-window)without cutoff time with cudf_frames. 
    **Cudf issue**   https://github.com/rapidsai/cudf/issues/6897

- [ ]  Related to **dtype selection**:

     Blocks feature:  Serialization/Writing to `csv`  for a cudf frame with list columns.  
     Cudf Issue: https://github.com/rapidsai/cudf/issues/6919

### Nice to have
- [x] Related to writing
This causes issues when encoding argument is provided (see issue: https://github.com/rapidsai/cudf/issues/7083 ) 

- [ ] Related to cutoffs as a cudf dataframe
We need dt.floor for this to be unlocked . ( issue: https://github.com/rapidsai/cudf/issues/7102 ) 

### Fixed 
- [x] Related to the primitives:
[Issue](https://github.com/rapidsai/cudf/issues/6921)  and [PR](https://github.com/rapidsai/cudf/pull/6973), (needed for null support to work for some comparison primitives)  

- [x] Related to **categorical** variables:

    Blocks feature:  Setting a  [last_time_index](https://docs.featuretools.com/en/v0.17.0/automated_feature_engineering/handling_time.html#setting-a-last-time-index) with cudf_frames  . 

    **Cudf issue:**   https://github.com/rapidsai/cudf/issues/6892 and  https://github.com/rapidsai/cudf/issues/6917

### CODE REVIEW

**Comment by VibhuJawa** [comment]
@kmax12 , Would it be possible to get an initial review in for this while we are waiting on the above mentioned cudf issues to be resolved to ensure that this approach is acceptable?

**Review by thehomebrewnerd — COMMENTED** [review]
@VibhuJawa Thank you for this contribution, and sorry for taking so long to get back to you. Anyway, I have taken a look through the PR, and I do not see any major issues here. It seems you are following the same approach that we have used for Dask and Koalas which is good.

At this point I do not have any significant feedback, but did notice a couple minor points, which I have commented on. 

I think next step is to merge in the latest changes from `main` and trigger a new run of the CI tests to make sure we are covering everything.

File: featuretools/tests/entityset_tests/test_es.py [hunk 1]
```diff
@@ -1395,6 +1485,7 @@ def test_same_index_values(index_df):
                                  dataframe=index_df,
                                  variable_types=vtypes)
 
+    print(index_df)
```
[the remark this answers is not in the document]
Comment (VibhuJawa) [reply featuretools/tests/entityset_tests/test_es.py]: Removed.

File: featuretools/tests/entityset_tests/test_entity.py [hunk 2]
```diff
@@ -15,6 +15,7 @@
 from featuretools.variable_types import find_variable_types
 
 ks = import_or_none('databricks.koalas')
+cudf = import_or_none('cudf')
```
[the remark this answers is not in the document]
Comment (VibhuJawa) [reply featuretools/tests/entityset_tests/test_entity.py:18]: There was an additional update regarding to list types needed, have added that, and now it is  now used.  See commit: https://github.com/alteryx/featuretools/pull/1246/commits/231ea04162b34a5b6c82d99e4e80adee6abe9e46

**Comment by VibhuJawa** [comment]
> I think next step is to merge in the latest changes from `main` and trigger a new run of the CI tests to make sure we are covering everything.

Thanks for the review @thehomebrewnerd .  Glad to know that i am on the right track. 

I have merged in main, addressed your reviews and updated the release notes but somehow the `CI tests` are still not being triggered.  Can you help me triage what is missing for that. 

Also, note that there are some known failures with `cudf`  as mentioned in the PR description that will need to be resolved for a successful test run.  We also will need a GPU CI for `cudf` as that needs GPU to work. I am not sure how you all run tests  `koalas` but if these are not run always with CI we can simply do the same with `cudf`.

**Comment by rwedge** [comment]
> I have merged in main, addressed your reviews and updated the release notes but somehow the `CI tests` are still not being triggered. Can you help me triage what is missing for that.

It looks like the tests are being run on the fork instead: https://github.com/VibhuJawa/featuretools/actions

Probably due to the choice of github action event type we use to trigger the tests 

> Also, note that there are some known failures with `cudf` as mentioned in the PR description that will need to be resolved for a successful test run. We also will need a GPU CI for `cudf` as that needs GPU to work. I am not sure how you all run tests for `koalas` but if these are not run always with CI we can simply skip them for `cudf`.

What we do with koalas currently is the "minimal" CI tests do not install koalas / spark and check the non-spark tests still pass.  The "optional" CI tests install koalas and run the koalas tests.

**Comment by VibhuJawa** [comment]
> Edit: created #1298 to track this issue

Thanks a lot for triaging and raising this.   

> What we do with koalas currently is the "minimal" CI tests do not install koalas / spark and check the non-spark tests still pass. The "optional" CI tests install koalas and run the koalas tests.

Gotcha, Thanks for the context around this. We can probably look into something similar for cudf once all the other issues are sorted out, i.e i have a local pass for all of the tests with `cudf`.

**Comment by VibhuJawa** [comment]
> I think next step is to merge in the latest changes from `main` and trigger a new run of the CI tests to make sure we are covering everything.

Just an FYI, while we are waiting on https://github.com/alteryx/featuretools/issues/1298 , I  synced my main and cudf_poc branches and we seem to have success on CI.  

See [RUN](https://github.com/VibhuJawa/featuretools/actions/runs/481190637
).

File: featuretools/entityset/serialize.py [hunk 3]
```diff
@@ -102,19 +105,38 @@ def write_entity_data(entity, path, format='csv', **kwargs):
             df = df.copy()
             columns = list(df.select_dtypes('object').columns)
             df[columns] = df[columns].astype(str)
-        df.to_csv(
```
[the remark this answers is not in the document]
Comment (VibhuJawa) [reply featuretools/entityset/serialize.py:105]: Resolved by commit https://github.com/alteryx/featuretools/pull/1246/commits/53f40ae182f43ba95e08e5511dbda5a6b7a9cd2d

**Comment by VibhuJawa** [comment]
So this PR currently fails due to lack of test coverage as it can't run code blocks (tests and mainline code)  as they will need cudf to be installed in the environment (which will essentially need GPUs).  


In an ideal world, we could have a GPU-CI running these tests, but it is currently difficult to set that up due to a lack of resources and infrastructure. 

As a stop-gap solution,  We can make the GPU CI run optional, which as an Nvidia employee, I can run locally on a schedule that you all recommend. I am happy to take on the maintenance burden of this if you guys are comfortable with this solution. Please let me know what you all think. 

CC: @thehomebrewnerd .

**Comment by thehomebrewnerd** [comment]
> As a stop-gap solution, We can make the GPU CI run optional, which as an Nvidia employee, I can run locally on a schedule that you all recommend. I am happy to take on the maintenance burden of this if you guys are comfortable with this solution. Please let me know what you all think.

@VibhuJawa I just realized we never responded to this. At this point, I think this is something we need to think about and explore a bit further. Definitely appreciate your willingness to take on the burden, but would be nice if we could identify an automated solution where that wouldn't be required.

File: nvidia_featuretools_license.md [hunk 4]
```diff
@@ -0,0 +1,64 @@
+The changes made to the following files in the https://github.com/alteryx/featuretools/pull/1246 are under the
```
Comment (gsheni) [inline nvidia_featuretools_license.md:1]: Can you remove this file?

Comment (VibhuJawa) [reply nvidia_featuretools_license.md]: Removed.

**Pull request closed without merging**